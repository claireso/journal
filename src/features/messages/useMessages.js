import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

const MessagesContext = React.createContext()

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

export const MessagesProvider = ({ children }) => {
  const [state, setState] = useState([])

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = () => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status) => (message) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
