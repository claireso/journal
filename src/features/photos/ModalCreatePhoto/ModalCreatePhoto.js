import { memo } from 'react'
import PropTypes from 'prop-types'

import { Heading2 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

const ModalCreatePhoto = ({ onSubmit, isProcessing }) => {
  return (
    <>
      <Heading2>Create a photo</Heading2>
      <FormPhoto onSubmit={onSubmit} isProcessing={isProcessing} />
    </>
  )
}

ModalCreatePhoto.propTypes = {
  onSubmit: PropTypes.func,
  isProcessing: PropTypes.bool
}

ModalCreatePhoto.defaultProps = {
  onSubmit: () => {},
  isProcessing: false
}

export default memo(ModalCreatePhoto)
