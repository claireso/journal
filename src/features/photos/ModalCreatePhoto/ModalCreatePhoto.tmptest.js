//@TODO restore test

// import { render, fireEvent, act } from '@testing-library/react'

// import ModalCreatePhoto from './ModalCreatePhoto'

// // import PhotosReducer from '../../reducer'
// import { PhotosProvider } from '../usePhotos'
// import { MessagesProvider } from '../../messages/useMessages'

// // const { PhotosProvider, INITIAL_STATE } = PhotosReducer

// // const { usePhotosReducer: originalUsePhotosReducer } = PhotosReducer

// describe('<Create />', () => {
//   const renderComponent = () => {
//     return render(
//       <MessagesProvider>
//         <PhotosProvider>
//           <ModalCreatePhoto />
//         </PhotosProvider>
//       </MessagesProvider>
//     )
//   }

//   // afterEach(() => {
//   //   PhotosReducer.usePhotosReducer = originalUsePhotosReducer
//   // })

//   test('should render component', () => {
//     const { container } = renderComponent()

//     expect(container.querySelector('input[name="color"]')).toBeNull()
//     expect(container).toMatchSnapshot()
//   })

//   test('should submit form', async () => {
//     const spySubmit = jest.fn()
//     // PhotosReducer.usePhotosReducer = () => [INITIAL_STATE, { createResource: spySubmit }]

//     const { container, getByLabelText } = renderComponent()

//     fireEvent.change(getByLabelText(/title/i), {
//       target: { value: 'Photo title' }
//     })
//     fireEvent.change(getByLabelText(/description/i), {
//       target: { value: 'Photo description' }
//     })

//     const inputFile = container.querySelector('input[type="file"]')

//     await act(async () => {
//       await fireEvent.change(inputFile, {
//         target: {
//           files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
//         }
//       })
//     })

//     fireEvent.change(getByLabelText(/position/i), {
//       target: { value: 'center' }
//     })

//     fireEvent.click(container.querySelector('input[type="submit"]'))

//     const expectedResult = new FormData(container.querySelector('form'))

//     expect(spySubmit).toHaveBeenCalledWith(expectedResult)
//   })

//   // test('should not submit form', async () => {
//   //   PhotosReducer.usePhotosReducer = () => [{ status: 'pending' }, {}]

//   //   const { container } = renderComponent()

//   //   expect(container.querySelector('input[type="submit"]')).toBeNull()
//   // })
// })
