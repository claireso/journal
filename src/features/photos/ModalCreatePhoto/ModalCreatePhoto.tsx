import { memo } from 'react'

import { Heading2 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

interface ModalCreatePhotoProps {
  onSubmit: (data: FormData) => void
  isProcessing?: boolean
}

const ModalCreatePhoto = ({ onSubmit, isProcessing = false }: ModalCreatePhotoProps) => {
  return (
    <>
      <Heading2>Create a photo</Heading2>
      <FormPhoto onSubmit={onSubmit} isProcessing={isProcessing} />
    </>
  )
}

export default memo(ModalCreatePhoto)
