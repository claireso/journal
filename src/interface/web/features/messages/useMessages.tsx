'use client'

import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

interface Message {
  key?: string
  status: 'default' | 'success' | 'error' | 'info'
  message: string
}

interface Actions {
  displaySuccessMessage: (data: Omit<Message, 'status'>) => void
  displayErrorMessage: (data: Omit<Message, 'status'>) => void
  closeMessage: (index: number) => void
  closeAllMessages: () => void
}

const INITIAL_STATE: Message[] = []

type TMessagesContext<T> = [T, React.Dispatch<React.SetStateAction<T>>]

const MessagesContext = React.createContext<TMessagesContext<Message[]>>([null!, null!])

interface MessagesProviderProps {
  children: React.ReactNode
}

export const MessagesProvider = ({ children }: MessagesProviderProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = (): [Message[], Actions] => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index: number) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status: Message['status']) => (message: Omit<Message, 'status'>) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message: Message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
