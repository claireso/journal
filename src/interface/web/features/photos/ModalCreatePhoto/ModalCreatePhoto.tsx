import { memo } from 'react'

import type { PhotoInsertDto } from '@dto'
import { Heading2 } from '@web/components/Headings'
import FormPhoto from '../FormPhoto'

interface ModalCreatePhotoProps {
  onSubmit: (data: PhotoInsertDto) => void
  isProcessing?: boolean
}

const ModalCreatePhoto = ({ onSubmit, isProcessing = false }: ModalCreatePhotoProps) => {
  return <FormPhoto<PhotoInsertDto> onSubmit={onSubmit} isProcessing={isProcessing} />
}

export default memo(ModalCreatePhoto)
