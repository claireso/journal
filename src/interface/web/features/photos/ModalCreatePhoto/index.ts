import ModalCreatePhoto from './ModalCreatePhoto'

export default ModalCreatePhoto
