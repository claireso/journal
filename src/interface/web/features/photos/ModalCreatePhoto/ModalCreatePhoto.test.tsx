import { screen, render, fireEvent, act, waitFor } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

import ModalCreatePhoto from './ModalCreatePhoto'

import { MessagesProvider } from '../../messages/useMessages'

describe('<ModalCreatePhoto />', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  })

  const renderComponent = (props = {}) => {
    return render(
      <QueryClientProvider client={queryClient}>
        <MessagesProvider>
          <ModalCreatePhoto onSubmit={jest.fn()} {...props} />
        </MessagesProvider>
      </QueryClientProvider>
    )
  }

  it('should render component', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it.skip('should create photo', async () => {
    const props = {
      onSubmit: jest.fn()
    }

    renderComponent(props)

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: 'Photo title' }
    })

    fireEvent.change(screen.getByLabelText(/description/i), {
      target: { value: 'Photo description' }
    })

    const inputFile = document.querySelector('input[type="file"]')

    if (!inputFile) {
      throw new Error('missing input in document')
    }

    await act(async () => {
      await fireEvent.change(inputFile, {
        target: {
          files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
        }
      })
    })

    await waitFor(() => expect(screen.getAllByTestId('preview')))

    fireEvent.change(screen.getByLabelText(/position/i), {
      target: { value: 'center' }
    })

    fireEvent.click(screen.getByTestId('submit'))

    const form = document.querySelector('form')

    if (!form) {
      throw new Error('missing form in document')
    }

    const formData = new FormData(form)

    expect(props.onSubmit).toHaveBeenCalledWith(formData)
  })
})
