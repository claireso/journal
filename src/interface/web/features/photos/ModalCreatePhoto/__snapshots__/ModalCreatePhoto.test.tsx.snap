// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalCreatePhoto /> should render component 1`] = `
<DocumentFragment>
  <form
    action=""
    enctype="multipart/form-data"
    method="POST"
  >
    <div
      class="styles__12zyxsm3 styles_layerStyle__12zyxsm0 styles_wrapper_inline_false__12zyxsm5 styles_layerStyle__12zyxsm2"
    >
      <label
        class="styles_layerStyle__17qmo0z0"
        for="title"
      >
        Title
      </label>
      <input
        class="styles_layerStyle__1d6x3p80"
        id="title"
        name="title"
        type="text"
        value=""
      />
    </div>
    <div
      class="styles__12zyxsm3 styles_layerStyle__12zyxsm0 styles_wrapper_inline_false__12zyxsm5 styles_layerStyle__12zyxsm2"
    >
      <label
        class="styles_layerStyle__17qmo0z0"
        for="description"
      >
        Description
      </label>
      <input
        class="styles_layerStyle__1d6x3p80"
        id="description"
        name="description"
        type="text"
        value=""
      />
    </div>
    <div
      class="styles__12zyxsm3 styles_layerStyle__12zyxsm0 styles_wrapper_inline_false__12zyxsm5 styles_layerStyle__12zyxsm2"
    >
      <label
        class="styles_layerStyle__17qmo0z0"
        for="file"
      >
        Photo
      </label>
      <div
        class="styles_layerStyle__xxgli80"
      >
        <div>
          <div
            class="styles_layerStyle__xxgli86"
          >
            <svg
              aria-hidden="true"
              class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_lg__19o76jud styles_layerStyle__19o76ju4 styles_icon_variant_primary__19o76juf styles_layerStyle__19o76ju6"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5m-13.5-9L12 3m0 0 4.5 4.5M12 3v13.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span
              class="styles_text__kgzpwem styles_text_size_base__kgzpweq styles_layerStyle__kgzpwe3 styles_text_color_base__kgzpwev styles_layerStyle__kgzpwe8 styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
            >
              Upload new photo
            </span>
            <span
              class="styles_text__kgzpwem styles_text_size_sm__kgzpwep styles_layerStyle__kgzpwe2 styles_text_color_neutral__kgzpwex styles_layerStyle__kgzpwea styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
            >
              (only image/jpeg, image/jpg)
            </span>
          </div>
          <input
            accept="image/jpeg,image/jpg"
            class="styles_layerStyle__xxgli87"
            name="file"
            required=""
            type="file"
          />
        </div>
      </div>
    </div>
    <div
      class="styles__12zyxsm3 styles_layerStyle__12zyxsm0 styles_wrapper_inline_true__12zyxsm4 styles_layerStyle__12zyxsm1"
    >
      <label
        class="styles_layerStyle__17qmo0z0"
        for="position"
      >
        Position
      </label>
      <div
        class="styles_layerStyle__1v4xt870"
      >
        <select
          class="styles_layerStyle__1v4xt871"
          id="position"
          name="position"
          required=""
        >
          <option
            selected=""
            value="left"
          >
            Left
          </option>
          <option
            value="center"
          >
            Center
          </option>
          <option
            value="right"
          >
            Right
          </option>
        </select>
        <div
          class="styles_layerStyle__1v4xt872"
        >
          <svg
            aria-hidden="true"
            class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_sm__19o76jub styles_layerStyle__19o76ju2 styles_icon_variant_default__19o76juh styles_layerStyle__19o76ju8"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m19.5 8.25-7.5 7.5-7.5-7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="styles_layerStyle__1g5jao10"
    >
      <button
        class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_primary__1mdc28jh styles_layerStyle__1mdc28j2 styles_button_size_lg__1mdc28jm styles_layerStyle__1mdc28j7"
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
</DocumentFragment>
`;
